python/services/dataservice/dmod/dataservice/rest_service.py

```diff
@@ -0,0 +1,144 @@
+)
+
+
+@app.exception_handler(ErrorResponseException)
+async def error_response_exception_handler(_: Request, exc: ErrorResponseException):
+    """Transforms ErrorResponseException's thrown during a request into a JSONResponse with status
+    code and error information from the ErrorResponseException's inner ErrorEnum member, `error`.
+
+    Example:
+    ```
+    from dmod.dataservice.errors import Error as ErrorEnum
+    from dmod.dataservice.exceptions import Error as ErrorResponseException
+
+    @app.get("/create")
+    def create_dataset_handler(name: str):
+        # ...
+        raise ErrorResponseException(ErrorEnum.DATASET_EXISTS)
+        # response (response status code is same as status):
+        # {
+        #     "type": "/errors/dataset_exists",
+        #     "title": "DATASET_EXISTS",
+        #     "status": 403,
+        #     "detail": "Dataset already exists"
+        # }
+        #
+        # or
+        # if detail not supplied, detail from ErrorEnum variant used
+        raise ErrorResponseException(ErrorEnum.DATASET_EXISTS, detail=f"A dataset already exists with name: {name}")
+    ```
+    """
+    return JSONResponse(
+        status_code=exc.error.status,
+        content=Error.from_error_enum(error_enum=exc.error, detail=exc.detail).dict(),
+    )
+
+
+@app.post("/create")
+async def create_dataset(
+    dataset_name: str,
+    data_category: DataCategory,
+    data_domain: DataDomain,
+) -> DatasetShortInfo:
+    """Create an empty dataset"""
+    ...
+
+
+@app.get("/get_object")
+async def get_object(dataset_name: str, object_path: Path) -> StreamingResponse:
+    """Download object from dataset."""
+    ...
+
+
+@app.put("/put_object")
+async def pub_object(
+    dataset_name: str,
+    object_name: Path,
+    object: UploadFile,
+    extract: bool = Query(
+        False,
+        description="If true, .tar.gz and .gz archive files will be extracted",
+    ),
+):
+    """Add a file or archive to an existing dataset. `.tar.gz` and `.gz` archive files will be
+    extracted if 'extract' parameter is true. Overwrites existing object with identical object_name.
+    """
+    ...
+
+
+@app.put("/put_objects")
+async def pub_objects(data: Annotated[DatasetPutObjectMultipleRequest, Depends()]):
+    """Add multiple files and add them to existing dataset. Overwrites existing objects with
+    identical object_names."""
+    ...
+
+
+@app.get("/list_objects")
+async def list_objects(
+    dataset_name: str, prefix: Optional[Path] = None, recursive: bool = True
+) -> List[DatasetObjectMetadata]:
+    """List a dataset's objects."""
+    ...
+
+
+@app.get("/query", response_model_exclude_none=True)
+async def query(dataset_name: str, q: List[QueryType]) -> DatasetQueryResponse:
+    """Inquire a dataset's metadata attributes."""
+    ...
+
+
+@app.delete("/delete")
+async def delete(dataset_name: str):
+    """Delete a dataset."""
+    ...
+
+
+@app.get("/search", response_model_exclude_none=True)
+async def search(
```

Comment:
We might need to add a `strict` parameter (or the inverse) to be able to control whether or not datasets with compatible formats should be considered.  It may need to be even more fine grained if we only want them considered when there are no matches having the exact format within the given `data_domain`.

We might also need to alter things to account for multiple matching datasets, including scenarios when these all have the exact format from the domain.

Reply:
Totally agree. Thanks for thinking of this and bringing it up! This api seems like it will be a little volatile until we can come up with search semantics that we are happy with. I wonder if we can get away with making all the fields in `DataDomain` and `DataCategory` optional or some sort of wildcard? Either way, I think we are going to have to think more about this.

Reply:
I think I may have clicked `Resolve` to that one by mistake earlier, so (at least for now) undoing that.

Reply:
Yeah I'm confused. I wrote "Nevermind..." in response to the prior resolved comment regarding the typo on `put_object`. Weird. Thanks for unresolving this and disregard the never mind.